source/particle/integrator/euler.cc

```diff
@@ -40,7 +40,13 @@ namespace aspect
                           "to the number of particles to advect. For some unknown reason they are different, "
                           "most likely something went wrong in the calling function."));
 
-        const bool geometry_has_periodic_boundary = (this->get_geometry_model().get_periodic_boundary_pairs().size() != 0);
+        const auto &cell = begin_particle->get_surrounding_cell();
+        bool at_periodic_boundary = false;
+        for (const auto &face_index: cell->face_indices())
+          if (cell->at_boundary(face_index))
+            if (cell->has_periodic_neighbor(face_index))
+              at_periodic_boundary = true;
```

Comment:
I think you can do even better, first by testing whether the geometry has periodic boundaries before starting the loop, and second by using `break` in the inner `if`:
```suggestion
        const auto &cell = begin_particle->get_surrounding_cell();
        bool at_periodic_boundary = false;
        if (this->get_geometry_model().get_periodic_boundary_pairs().size() != 0)
          for (const auto &face_index: cell->face_indices())
            if (cell->at_boundary(face_index))
              if (cell->has_periodic_neighbor(face_index))
              {
                at_periodic_boundary = true;
                break;
              }
```